### PR TITLE
Bug 1919336: Check if datastore belongs to a datastore cluster

### PR DESCRIPTION
**Pull request by gnufied**
fixes https://bugzilla.redhat.com/show_bug.cgi?id=1919336

Output after the change:

```
I0208 17:00:29.199915  150092 operator.go:251] CheckDefaultDatastore failed: defaultDatastore "team-ssd-intel" in vSphere configuration: datastore team-ssd-intel is part of intel-ds-cluster datastore cluster

```

### CODE REVIEW

File: pkg/check/datastore.go [hunk 1]
```diff
@@ -226,16 +234,111 @@ func getPolicy(ctx *CheckContext, name string) ([]types.BasePbmProfile, error) {
...
+	finder = find.NewFinder(ctx.VMClient, false)
+	finder.SetDatacenter(matchingDC)
+	tctx, cancel = context.WithTimeout(ctx.Context, *Timeout)
+	defer cancel()
+	ds, err := finder.Datastore(tctx, dataStoreName)
+	if err != nil {
+		klog.Errorf("error getting datastore %s: %v", dataStoreName, err)
+		return nil
+	}
+
```
Comment (jsafrane) [inline pkg/check/datastore.go:277]: This is used to get DC/DS elsewhere. Is it wrong?
https://github.com/openshift/vsphere-problem-detector/blob/ffe5e99596fb3c1a363c96ee057614bbaf997a15/pkg/check/folder.go#L22-L34

File: pkg/check/datastore.go [hunk 2]
```diff
@@ -226,16 +234,111 @@ func getPolicy(ctx *CheckContext, name string) ([]types.BasePbmProfile, error) {
...
+			if err != nil {
+				klog.Errorf("fetching datastore %s failed: %v", child.String(), err)
+				continue
+			}
+			if tDS.Summary.Url == dsMo.Summary.Url {
+				return fmt.Errorf("datastore %s is part of %s datastore cluster", tDS.Summary.Name, ds.Summary.Name)
+			}
+		}
+	}
+	klog.V(2).Infof("Checked datastore %s for SRDS - no problems found", dataStoreName)
```
Comment (jsafrane) [inline pkg/check/datastore.go:325]: This is called for every StorageClass  and might be too noisy.

File: pkg/check/datastore.go [hunk 3]
```diff
@@ -226,16 +234,111 @@ func getPolicy(ctx *CheckContext, name string) ([]types.BasePbmProfile, error) {
...
+	err = v.Destroy(tctx)
+	if err != nil {
+		klog.Errorf("error destroying view: %+v", err)
+		return nil
+	}
+	for _, ds := range content {
+		for _, child := range ds.Folder.ChildEntity {
+			tDS, err := getDatastore(ctx, child)
+			if err != nil {
+				klog.Errorf("fetching datastore %s failed: %v", child.String(), err)
```
Comment (jsafrane) [inline pkg/check/datastore.go:317]: Why is the error ignored? A comment would be nice.

File: pkg/check/datastore.go [hunk 4]
```diff
@@ -226,16 +234,111 @@ func getPolicy(ctx *CheckContext, name string) ([]types.BasePbmProfile, error) {
...
+	if err != nil {
+		klog.Errorf("error listing datastore cluster: %+v", err)
+		return nil
+	}
+	var content []mo.StoragePod
+	tctx, cancel = context.WithTimeout(ctx.Context, *Timeout)
+	defer cancel()
+	err = v.Retrieve(tctx, kind, []string{SummaryProperty, "childEntity"}, &content)
+	if err != nil {
+		klog.Errorf("error retrieving datastore cluster properties: %+v", err)
```
Comment (jsafrane) [inline pkg/check/datastore.go:305]: should be `v` destroyed on error?

File: pkg/check/datastore.go [hunk 5]
```diff
@@ -226,16 +234,111 @@ func getPolicy(ctx *CheckContext, name string) ([]types.BasePbmProfile, error) {
...
+		return nil
+	}
+
+	var dsMo mo.Datastore
+	pc := property.DefaultCollector(matchingDC.Client())
+	properties := []string{DatastoreInfoProperty, SummaryProperty}
+	tctx, cancel = context.WithTimeout(ctx.Context, *Timeout)
+	defer cancel()
+	err = pc.RetrieveOne(tctx, ds.Reference(), properties, &dsMo)
+
```
Comment (jsafrane) [inline pkg/check/datastore.go:284]: nit: extra empty line

File: pkg/check/datastore.go [hunk 6]
```diff
@@ -226,16 +234,111 @@ func getPolicy(ctx *CheckContext, name string) ([]types.BasePbmProfile, error) {
...
 	if err := checkVolumeName(volumeName); err != nil {
 		return fmt.Errorf("datastore %s: %s", dsName, err)
 	}
+	if err := checkForDatastoreCluster(ctx, dsName); err != nil {
+		return err
+	}
 	return nil
 }
 
+func checkForDatastoreCluster(ctx *CheckContext, dataStoreName string) error {
```
Comment (jsafrane) [inline pkg/check/datastore.go:250]: Perhaps for a subsequent PR: This is called for every StorageClass and loads non-trivial amount of data from vSphere. Would it be possible to call it only once per `dataStorageName`? I.e. with a cache _somewhere_ in `CheckStorageClasses`? It's not that easy, because there is CheckDefaultDatastore too...

File: pkg/check/datastore.go (hunk 3 above)
Comment (gnufied) [reply pkg/check/datastore.go:317]: I could add a comment. It is being ignored because - it is possible that the OCP user may not have permission to fetch datastore that is not being used in OCP.

File: pkg/check/datastore.go [hunk 7]
```diff
@@ -6,17 +6,25 @@ import (
...
 
 const (
 	dsParameter            = "datastore"
 	storagePolicyParameter = "storagepolicyname"
+	dataCenterType         = "Datacenter"
+	DatastoreInfoProperty  = "info"
+	SummaryProperty        = "summary"
 )
 
 // CheckStorageClasses tests that datastore name in all StorageClasses in the cluster is short enough.
```
[the remark this answers is not in the document]
Comment (gnufied) [reply pkg/check/datastore.go:30]: When we look at a PV source - we only the datastore it is present in. it does not tells us which datacenter the vmdk file belongs to. So theoretically it is possible that - there could be more than one datastore with same name in different datacenters and hence we can't uniquely identify datastore with just the VMDK path.

We could perhaps do a weaker version of the check though by assuming the datacenter is the one specified in workspace. This assumption may not always hold true. Fortunately OCP does not support multiple zones but one can still statically provision PVs with nodeAffinity and such (although I would think the nodes are not zone labelled in vsphere).

[the remark this answers is not in the document]
Comment (gnufied) [reply pkg/check/datastore.go:30]: I left similar comment to missing VMDKs check - https://bugzilla.redhat.com/show_bug.cgi?id=1904482#c1

File: pkg/check/datastore.go (hunk 1 above)
Comment (gnufied) [reply pkg/check/datastore.go:277]: the logic to fetch DS is same. I think I used somewhat different logic to fetch datacenter. I will make util function for that.

File: pkg/check/datastore.go (hunk 4 above)
Comment (gnufied) [reply pkg/check/datastore.go:300]: using defer now.

File: pkg/check/datastore.go (hunk 3 above)
Comment (gnufied) [reply pkg/check/datastore.go:313]: added

File: pkg/check/datastore.go (hunk 2 above)
Comment (gnufied) [reply pkg/check/datastore.go:321]: increased the log level.

File: pkg/check/datastore.go (hunk 6 above)
Comment (gnufied) [reply pkg/check/datastore.go:254]: Yes we could. I think it might be worth caching the information within the same run. but once all the checks are run -we can automatically expire the cache, so as they aren't reused in next run.

File: pkg/check/datastore.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (jsafrane) [reply pkg/check/datastore.go:33]: Ack

File: pkg/check/datastore.go [hunk 8]
```diff
@@ -59,7 +65,7 @@ func CheckStorageClasses(ctx *CheckContext) error {
 			}
 		}
 	}
-	klog.V(2).Infof("CheckStorageClasses checked %d storage classes, %d problems found", len(scs), len(errs))
+	klog.V(4).Infof("CheckStorageClasses checked %d storage classes, %d problems found", len(scs), len(errs))
```
Comment (jsafrane) [inline pkg/check/datastore.go:68]: Why this change? This log is useful.

File: pkg/check/datastore.go (hunk 2 above)
Comment (jsafrane) [reply pkg/check/datastore.go:321]: You increased it at wrong place, see above

File: pkg/check/datastore.go (hunk 1 above)
Comment (jsafrane) [reply pkg/check/datastore.go]: What `checkForDatastoreCluster` / `CheckFolderPermissions` need is actually `getDatastore` (with `getDatacenter` inside) - fetching the datastore is copied over there.

Comment (gnufied) [reply pkg/check/datastore.go]: I refactored this code bit more. But in `checkForDatastoreCluster` function we are also fetching datastore summary/info so as we can get datastore URL. This field is not available in `object.Datastore` value.

File: pkg/check/datastore.go (hunk 2 above)
Comment (gnufied) [reply pkg/check/datastore.go]: yikes. fixed.

**Comment by gnufied** [comment]
@jsafrane fixed the comments. PTAL

File: pkg/check/datastore.go (hunk 1 above)
Comment (jsafrane) [reply pkg/check/datastore.go]: Thanks!
